src/Arcus.Messaging.Abstractions/MessageHandling/MessageHandler.cs

```diff
@@ -106,19 +106,29 @@ private static Type[] GetServiceRegistrationTypes(IServiceProvider serviceProvid
 
         private static IEnumerable<Type> GetServiceRegistrationTypesFromMicrosoftExtensions(IServiceProvider serviceProvider, ILogger logger)
         {
-            object engine = 
-                serviceProvider.GetType().GetProperty("Engine")?.GetValue(serviceProvider) 
-                ?? serviceProvider.GetType().GetField("_engine", BindingFlags.NonPublic | BindingFlags.Instance)?.GetValue(serviceProvider);
+            object engine =
```

Comment:
You'll probably wonder why we have this reflection code in the `MessageHandler` file.
This is bc we were faced with a problem when implementing the message routing. The `IServceProvider` doesn't support open generic interfaces and since our `IMessageHandler<,>` is such an open interface, there is no build-in way to extract all service registrations with a generic message type.
This message handler class will do this for us. It searches for the 'engine' and 'call site factory' on the service provider and extracts all the message handler interfaces. Later on, we can determine if there exists and message handler that can process the incoming message.

We can off course discuss this later, or come up with something else. For now, this was the best we could do. Unfortunately.

Reply:
I was indeed wondering,  It might be good to have a short call on this once upon a time to explain this :)